avogadro/qtplugins/surfaces/surfaces.cpp

```diff
@@ -183,69 +186,149 @@ void Surfaces::calculateSurface()
-  m_cube->setLimits(*m_molecule, m_dialog->resolution(), padding);
-
-  auto neighborPerceiver = Core::NeighborPerceiver(
-    m_molecule->atomPositions3d(), 2.0 * max_radius
-  );
-  Vector3i cubeSize = m_cube->dimensions();
-
-  std::vector<Vector3i> *indices = new std::vector<Vector3i>;
-  indices->reserve(cubeSize[0] * cubeSize[1] * cubeSize[2]);
-  for (size_t zi = 0; zi < cubeSize[2]; zi++)
-    for (size_t yi = 0; yi < cubeSize[1]; yi++)
-      for (size_t xi = 0; xi < cubeSize[0]; xi++)
-        indices->emplace_back(xi, yi, zi);
-
-  const float res = m_dialog->resolution();
-  const Vector3 min = m_cube->min();
-  const float mdist = probeRadius;
-
-  thread_local Array<Index> *neighbors = nullptr;
-  QFuture future = QtConcurrent::map(*indices, [=](Vector3i &in) {
-    Vector3 pos(in(0), in(1), in(2));
-    pos += Vector3(0.5, 0.5, 0.5);
-    pos *= res;
-    pos += min;
-
-    double minDistance = mdist;
-    if (neighbors == nullptr)
-      neighbors = new Array<Index>;
-    neighborPerceiver.getNeighborsInclusiveInPlace(*neighbors, pos);
-    for (Index neighbor: *neighbors) {
-      Vector3 neighborPos = m_molecule->atomPosition3d(neighbor);
-      double distance = (neighborPos - pos).norm();
-      distance -= radii[neighbor] + radiusOffset;
-      minDistance = std::min(minDistance, distance);
+    // first, make a list of all atom positions and radii
+    Array<Vector3> atomPositions = m_molecule->atomPositions3d();
+    std::vector<std::pair<Vector3, double>> *atoms =
+      new std::vector<std::pair<Vector3, double>>(m_molecule->atomCount());
+    double max_radius = probeRadius;
+    for (size_t i = 0; i < atoms->size(); i++) {
+      (*atoms)[i].first = atomPositions[i];
+      (*atoms)[i].second = Core::Elements::radiusVDW(m_molecule->atomicNumber(i)) + probeRadius;
+      if ((*atoms)[i].second > max_radius)
+        max_radius = (*atoms)[i].second;
     }
 
-    m_cube->setValue(in(0), in(1), in(2), minDistance);
+    double padding = max_radius + probeRadius;
+    m_cube->setLimits(*m_molecule, m_dialog->resolution(), padding);
+    Vector3i size = m_cube->dimensions();
+    for (int z = 0; z < size(2); z++)
+      for (int y = 0; y < size(1); y++)
+        for (int x = 0; x < size(0); x++)
+          m_cube->setValue(x, y, z, -1.0f);
+
+    const float res = m_dialog->resolution();
+    const Vector3 min = m_cube->min();
+    const float mdist = probeRadius;
+
+    // then, for each atom, set cubes around it up to a certain radius
+    QFuture innerFuture = QtConcurrent::map(*atoms, [=](std::pair<Vector3, double> &in) {
+      double startPosZ = in.first(2) - in.second;
+      double endPosZ = in.first(2) + in.second;
+      int startIndexZ = (startPosZ - min(2)) / res;
+      int endIndexZ = (endPosZ - min(2)) / res + 1;
+      for (int indexZ = startIndexZ; indexZ < endIndexZ; indexZ++) {
+        double posZ = indexZ * res + min(2);
+        double radiusZsq = pow(in.second, 2) - pow(posZ - in.first(2), 2);
```

Comment:
I'm pretty sure that `in.second**in.second` is faster than `std:pow()`

Sometimes I've defined a macro:
`#define SQUARE(x) ((x)*(x))`

Reply:
I believe the fastest way to do this on all compilers is to define an inline function, as in some cases a macro could lead to its arguments being evaluated twice, and `pow(x, 2)` is optimized into a register multiply on GCC (`-O1`) and Clang (`-O1`), but apparently not MSVC.